### ISSUE TITLE
rkyv 0.7 build fails on riscv32 with floating point

### ISSUE THREAD

**Issue by mcclure** [rkyv/rkyv#469]
I am using rkyv in a configuration where I am using it to create rkyv serialized files in my build.rs, then include them in my executable with include_bytes. I am doing this to reduce memory/footprint on an embedded platform. The embedded platform is slightly unusual: `riscv32imafdc`¹. In my testing, I *believe* I see rkyv succeeds with the more common `riscv32imac` platform. On Discord, the rkyv maintainer said the problem was 0.7 has a Rust MSRV which bars automatic discovery of atomic support, but volunteered to manually add support for my platform to the 0.7 build.rs if it would be helpful. Because my attempt to switch to `main` branch/0.8-pre was unsuccessful (#468) , I would like to request this support be added.

### Repro

Check out [this example](https://github.com/mcclure/pocket-riscv-rs-bug/tree/z_bug_rkyv32) (branch `z_bug_rkyv32`). This is a fork of my actual project. Follow the "First time" and "After" instructions in `run.txt` (this is not doing anything unusual, basically you will need to update git submodules, tell rustup to install riscv support, and then `make` just does a normal cargo build with nightly). It will fail with:

```
mcc@Anthy:~/work/r/v/riscv-pocket-app$ make
cargo +nightly build  --release
   Compiling rkyv v0.7.43
   Compiling minibreak v0.1.0 (/home/mcc/work/r/v/riscv-pocket-app)
error[E0432]: unresolved imports `core::sync::atomic::AtomicI64`, `core::sync::atomic::AtomicU64`
 --> /home/mcc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.7.43/src/impls/core/primitive.rs:8:26
  |
8 | use core::sync::atomic::{AtomicI64, AtomicU64};
  |                          ^^^^^^^^^  ^^^^^^^^^ no `AtomicU64` in `sync::atomic`
  |                          |
  |                          no `AtomicI64` in `sync::atomic`
  |
help: a similar name exists in the module
  |
8 | use core::sync::atomic::{AtomicI32, AtomicU64};
  |                          ~~~~~~~~~
help: a similar name exists in the module
  |
8 | use core::sync::atomic::{AtomicI64, AtomicU32};
  |                                     ~~~~~~~~~

error[E0432]: unresolved imports `core::sync::atomic::AtomicI64`, `core::sync::atomic::AtomicU64`
 --> /home/mcc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.7.43/src/with/atomic.rs:9:26
  |
9 | use core::sync::atomic::{AtomicI64, AtomicU64};
  |                          ^^^^^^^^^  ^^^^^^^^^ no `AtomicU64` in `sync::atomic`
  |                          |
  |                          no `AtomicI64` in `sync::atomic`
  |
help: a similar name exists in the module
  |
9 | use core::sync::atomic::{AtomicI32, AtomicU64};
  |                          ~~~~~~~~~
help: a similar name exists in the module
  |
9 | use core::sync::atomic::{AtomicI64, AtomicU32};
  |                                     ~~~~~~~~~

For more information about this error, try `rustc --explain E0432`.
error: could not compile `rkyv` (lib) due to 2 previous errors
make: *** [Makefile:10: main] Error 101
```

By comparison, [this example](https://github.com/mcclure/pocket-riscv-rs-bug/tree/z_bug_rkyv32i) (branch z_bug_rkyv32i) uses the same rkyv configuration but targets the more common `riscv32imac` platform variant, and *this* example builds successfully (this example does *not* require nightly).

I did a third test (can supply code if helpful) attempting to target `riscv32imafc`, which ended inconclusively due to unrelated failures, but I *believe* `riscv32imafc` is seeing the same failures as `riscv32imafdc`.

Based on my limited understanding of the problem, if you hardcode `riscv32imafc` and `riscv32imafdc` to only support 32-bit atomics (on a casual read of the [riscv atomics extension](https://five-embeddev.com/riscv-isa-manual/latest/a.html), 32-bit riscv with atomics will *never* support 64-bit atomics), these two platforms will begin working.

### "Expected behavior"

A build of a project including stable rkyv should work on the `riscv32imafdc` and `riscv32imafc` platforms (although, of these two, riscv32imafdc is the only one I personally am using). 

---
¹ This may be more information than you require: RISCV is a "open hardware" chip ISA. The unusual platform names are due to it having an "extension" system. RISCV is effectively a family of chips where individual chips are identified by `riscv`, followed by `32` or `64` for the word size, followed by a series of one-letter "[extensions](https://en.wikichip.org/wiki/risc-v/standard_extensions)", each of which may be present or absent. This creates the possibility for dozens of combinations, but in practice only some combinations are used: `riscv32imac` and `riscv64imafdc` (the latter by convention shortened to `riscv64gc`) are by far the most common. In English, this means that currently most users either select a basic 32-bit integer configuration with multiplication and atomics, or a 64-bit configuration which adds floating point. (For my use case, I am targeting a chip synthesized for FPGA and running on the Analogue Pocket video game system, which is why we want 32-bit and floating point simultaneously.) In Rust, there is a ["tiered" support system](https://doc.rust-lang.org/nightly/rustc/platform-support.html) for platforms; `riscv32imac`, `riscv32imafc` (imac with 32-bit floating point *only*, whereas my configuration adds 64-bit floating point) and `riscv32gc` are all listed as "tier 2" platforms, whereas `riscv32imafdc` (32-bit pointers with 64-bit floating point) is "tier 3". Per the documentation, this means rustup will vend corelib binaries for the tier 2 platforms but not the tier 3 platforms (although in practice, I can get `rustup target add` to work with `riscv32imac` but *not* `riscv32imafc`). The upshot of all of this is that your current level of support is "reasonable", since `riscv32imac` can be (1) actually purchased as a real-world microchip and (2) easily targeted using official Rust project binaries without doing anything strange, whereas as far as I can tell neither of these is true for the floating point versions you do not currently support. However, without either `riscv32imafdc` support in 0.7 build.rs or a working no_std 0.8, my project is blocked.

**Comment by djkoloski** [comment]
This should be fixed as of 0.7.44, which adds additional `build.rs` detection for `riscv32imafdc`. You may need to update `rend` to 0.4.2 and `bytecheck` to 0.6.12, so a `cargo update` may be required.

**Comment by mcclure** [comment]
Can confirm, newest 0.7 can create archives in build.rs on PC and load them on the riscv target. Because the RKYV based system is more efficient than the one it replaces, this has almost doubled my project framerate. Thanks!
